### PR TITLE
Use TiKV

### PR DESCRIPTION
**Pull request by cherrypiejam**
Modified `singlevm`, `admin_fstools`, and `multivm` to use TiKV

### CODE REVIEW

File: snapfaas/bins/singlevm/main.rs [hunk 1]
```diff
@@ -335,7 +356,20 @@ fn main() {
     eprintln!("***********************************************");
 
     // Shutdown the vm and exit
-    eprintln!("Shutting down vm..."); 
+    eprintln!("Shutting down vm...");
     drop(vm);
     unlink_unix_sockets();
 }
+
+struct FSWrapper(Box<dyn snapfaas::fs::BackingStore>);
```
Comment (tan-yue) [inline snapfaas/bins/singlevm/main.rs:364]: Can we put this in, say, src/fs/wrapper.rs?

**Review by tan-yue — CHANGES_REQUESTED** [review]

File: snapfaas/bins/singlevm/main.rs (hunk 1 above)
Comment (tan-yue) [reply snapfaas/bins/singlevm/main.rs:364]: You are right.

Comment (cherrypiejam) [reply snapfaas/bins/singlevm/main.rs:364]: AsRef is a trait, so `impl<T: AsRef<B>, B: BackingStore> BackingStore for T`. The problem is that `B` isn't used anywhere, making it an unconstrained type parameter. Is there a workaround for this?

File: snapfaas/bins/admin_fstools/main.rs [hunk 2]
```diff
@@ -27,9 +27,29 @@ pub fn main() -> std::io::Result<()> {
                 .args(&["bootstrap", "update_fsutil", "update_python", "faceted-list", "list", "read", "blob", "mkdir", "delete"])
                 .required(true),
         )
+        .arg(
+            Arg::with_name("tikv")
+                .long("tikv")
+                .value_name("TIKV")
+                .takes_value(true)
+                .required(false)
+                .help("Use TiVK as the backing store.")
```
Comment (alevy) [inline snapfaas/bins/admin_fstools/main.rs:36]: In all cases (singlevm, admin_fstools, multivm) it's totally unclear what this new argument is meant to take. Presumably it's the list of tikv proxies? In what format? etc.

Comment (cherrypiejam) [reply snapfaas/bins/admin_fstools/main.rs]: Fixed. Previously it only takes one tikv proxy, and now it takes a list of them.

File: snapfaas/bins/admin_fstools/main.rs [hunk 3]
```diff
@@ -27,9 +27,29 @@ pub fn main() -> std::io::Result<()> {
...
+                .long("tikv")
+                .takes_value(true)
+                .required(false)
+                .help("One or more addresses of TiKV placement driver, separated by space.")
+        )
         .get_matches();
 
-    let fs = snapfaas::fs::FS::new(&*snapfaas::fs::lmdb::DBENV);
+    let fs: FS<Box<dyn BackingStore>> = FS::new(
+        match matches.value_of("tikv proxies").map(|ts| ts.split_whitespace().into_iter().collect()) {
```
Comment (tan-yue) [inline snapfaas/bins/admin_fstools/main.rs:41]: There is a more idiomatic way to do this https://docs.rs/clap/2.33.0/clap/struct.Arg.html#method.multiple and potentially with https://docs.rs/clap/2.33.0/clap/struct.Arg.html#method.use_delimiter

Comment (tan-yue) [reply snapfaas/bins/admin_fstools/main.rs:41]: I think what we should spend some time doing is to bump clap version to 4.x.x
The current 2.33 is very old. For now, I'm gonna just merge this.

**Review by tan-yue — APPROVED** [review]

File: snapfaas/bins/admin_fstools/main.rs (hunk 3 above)
Comment (cherrypiejam) [reply snapfaas/bins/admin_fstools/main.rs:41]: Got it. Sounds good

**Pull request merged**